kernel/arch/dreamcast/include/dc/g2bus.h

```diff
@@ -31,50 +33,165 @@
+
+/** \brief  Initialize DMA support.
+
+    This function sets up the DMA support for transfers to/from the G2 Bus.
 
-   Update : looks like there is a formula, mode = 3 + shchn
+    \retval 0               On success (no error conditions defined).
+*/
+int g2_dma_init(void);
+
+/** \brief  Shutdown DMA support. */
+void g2_dma_shutdown(void);
+
+/** \brief  G2 context
+
+    A G2 context containing the states of IRQs and G2 DMA. This struct
+    is used in with g2_lock() and g2_unlock().
+*/
+typedef struct { 
+    uint32_t irq_state;    /** \brief IRQ state when entering a G2 critical block */
+} g2_ctx_t;
+
+/* Internal constants */
+/** \cond */ 
+#define SPU_G2DMA_SUSPEND     (*((vuint32 *)0xa05f781C))
+#define BBA_G2DMA_SUSPEND     (*((vuint32 *)0xa05f783C))
+#define CH2_G2DMA_SUSPEND     (*((vuint32 *)0xa05f785C))
+/** \endcond */
+
+/** \brief  Disable IRQs and G2 DMA
+
+    This function makes the following g2_read_*()/g2_write_*() functions atomic 
+    by disabling IRQs and G2 DMA and storing their states. Pass the context 
+    created by this function to g2_unlock() to re-enable IRQs and G2 DMA.
 
+    \return                 The context containing the IRQ and G2 DMA states.
 */
+static inline g2_ctx_t g2_lock(void) {
+    g2_ctx_t ctx;
 
-/* We use sh channel 3 here to avoid conflicts with the PVR. */
-#define SPU_DMA_MODE   6 /* should we use 6 instead, so that the formula is 3+shchn ?
-6 works too, so ... */
-#define SPU_DMA_G2CHN  0
-#define SPU_DMA_SHCHN  3
+    ctx.irq_state = irq_disable();
+
+    /* Suspend any G2 DMA */
+    SPU_G2DMA_SUSPEND = 1;
+    BBA_G2DMA_SUSPEND = 1;
+    CH2_G2DMA_SUSPEND = 1;
+
+    while(FIFO_STATUS & FIFO_SH4);
+
+    return ctx;
+}
+
+/** \brief  Enable IRQs and G2 DMA
+
+    This function restores the IRQ and G2 DMA states using the context value
+    generated by g2_lock().
+
+    \param  ctx             The context containing IRQ and G2 DMA states.
+*/
+static inline void g2_unlock(g2_ctx_t ctx) {
+    /* Restore suspended G2 DMA */
+    SPU_G2DMA_SUSPEND = 0;
+    BBA_G2DMA_SUSPEND = 0;
+    CH2_G2DMA_SUSPEND = 0;
 
-/* For BBA : sh channel 1 (doesn't seem used) and g2 channel 1 to no conflict with SPU */
-#define BBA_DMA_MODE   4
-#define BBA_DMA_G2CHN  1
-#define BBA_DMA_SHCHN  1
+    irq_restore(ctx.irq_state);
+}
 
-/* For BBA2 : sh channel 0 (doesn't seem used) and g2 channel 2 to no conflict with SPU */
-/* This is a second DMA channels used for the BBA, just for fun and see if we can initiate
-   two DMA transfers with the BBA concurently. */
-#define BBA_DMA2_MODE   3
-#define BBA_DMA2_G2CHN  2
-#define BBA_DMA2_SHCHN  0
 
-typedef void (*g2_dma_callback_t)(ptr_t data);
-int g2_dma_transfer(void *from, void * dest, uint32 length, int block,
-                    g2_dma_callback_t callback, ptr_t cbdata,
-                    uint32 dir, uint32 mode, uint32 g2chn, uint32 sh4chn);
+#undef SPU_G2DMA_SUSPEND
+#undef BBA_G2DMA_SUSPEND
+#undef CH2_G2DMA_SUSPEND
```

Comment:
If these are renamed up above and documented, then I don't think it's necessary to undef them.

Reply:
Internal use and they aren't meant for user-code use.